src/west/manifest.py

```diff
@@ -537,14 +573,72 @@ def _load_project(self, pdata):
         except ValueError as ve:
             self._malformed(ve.args[0])
 
+        import_files = self._imported_files(project, import_key)
+
         # The name "manifest" cannot be used as a project name; it
         # is reserved to refer to the manifest repository itself
         # (e.g. from "west list"). Note that this has not always
         # been enforced, but it is part of the documentation.
         if project.name == 'manifest':
             self._malformed('no project can be named "manifest"')
 
-        return project
+        return project, import_files
+
+    def _imported_files(self, project, import_key):
+        ktype = type(import_key)
+
+        if not import_key or (isinstance(project, ManifestProject) and
+                              ktype == bool):
+            return []
+
+        if ktype == bool:
+            ret = os.path.join(project.abspath, _WEST_YML)
+            if not os.path.isfile(ret):
+                _file_not_found(ret)
+            return [ret]
+        elif ktype == str:
+            path = os.path.join(project.abspath, import_key)
+            if os.path.isfile(path):
+                # Take the user's word for it, even if that doesn't end in
+                # something from _YML_EXTS.
+                return [path]
+            elif os.path.isdir(path):
+                # Find the YAML files in the directory. It's OK if there
+                # aren't any.
+                ret = []
+                for f in os.listdir(path):
+                    if os.path.splitext(f) in _YML_EXTS:
+                        ret.append(f)
+                return ret
+            else:
+                _file_not_found(f)
+        elif ktype == list:
+            return list(itertools.chain(self._imported_files(project, k)
+                                        for k in import_key))
+        elif ktype == dict:     # TODO
+            raise NotImplementedError('import: <map> is unimplemented')
+        else:
+            self._malformed('{}: invalid import key {} (type {})'.
+                            format(project.name, import_key, ktype))
+
+    def _import(self, project, file):
+        try:
+            data = project.file_contents_at(file)
+        except subprocess.CalledProcessError:
+            # Fall back on the importer.
+            data = self._importer(project, file, project.revision)
+
+        m = Manifest(source_data=data, manifest_path=project.abspath,
```

Comment:
There's no loop detection here yet, but that's indeed where it should go.

(It can't go into the importer for west update, because it needs to be checked every time we import.)